File: src/main/kotlin/net/dontdrinkandroot/gitki/wicket/model/FileSizeStringModel.java
package net.dontdrinkandroot.gitki.wicket.model;

import net.dontdrinkandroot.wicket.model.AbstractChainedReadonlyModel;
import org.apache.wicket.model.IModel;

import java.text.DecimalFormat;

/**
 * @author Philip Washington Sorst <philip@sorst.net>
 */
public class FileSizeStringModel extends AbstractChainedReadonlyModel<Long, String>
{
    public FileSizeStringModel(IModel<? extends Long> parent) {
        super(parent);
    }

    @Override
    public String getObject() {
        Long size = this.getParentObject();

        if (size <= 0)
            return "0";
        final String[] units = new String[]{"B", "kB", "MB", "GB", "TB"};
        int digitGroups = (int) (Math.log10(size) / Math.log10(1024));

        return new DecimalFormat("###0.#").format(size / Math.pow(1024, digitGroups)) + " " + units[digitGroups];
    }
}

File: src/main/kotlin/net/dontdrinkandroot/gitki/wicket/model/FileSizeStringModel.kt
package net.dontdrinkandroot.gitki.wicket.model

import net.dontdrinkandroot.wicket.model.AbstractChainedModel
import org.apache.wicket.model.IModel
import java.text.DecimalFormat
import kotlin.math.log10
import kotlin.math.pow

class FileSizeStringModel(parent: IModel<out Long>) : AbstractChainedModel<Long, String>(parent) {

    override fun getValue(parentValue: Long?): String? = parentValue?.let {
        if (it <= 0) return "0"
        val units = arrayOf("B", "kB", "MB", "GB", "TB")
        val digitGroups = (log10(it.toDouble()) / log10(1024.0)).toInt()
        return DecimalFormat("###0.#").format(it / 1024.0.pow(digitGroups.toDouble())) + " " + units[digitGroups]
    }
}